src/hotspot/share/services/rsswatch.cpp

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2023, Red Hat, Inc. All rights reserved.
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "logging/log.hpp"
+#include "runtime/java.hpp"
+#include "runtime/os.hpp"
+#include "runtime/task.hpp"
+#include "services/rsswatch.hpp"
+#include "utilities/parseInteger.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+static void check_rss(size_t limit) {
+  const size_t rss = os::get_rss();
+  log_trace(os, rss)("Rss=%zu", rss);
+  if (rss >= limit) {
+    fatal("Resident Set Size (%zu bytes) reached RssLimit (%zu bytes).", rss, limit);
+  }
+}
+
+class RssAbsoluteLimitTask : public PeriodicTask {
+  const size_t _limit;
+public:
+  RssAbsoluteLimitTask(size_t limit, unsigned interval_ms) :
+    PeriodicTask(interval_ms), _limit(limit) {
+    log_info(os, rss)("RssWatcher task: interval=%ums, limit=" PROPERFMT,
+                      interval(), PROPERFMTARGS(_limit));
+  }
+  void task() override {
+    check_rss(_limit);
+  }
+};
+
+class RssRelativeLimitTask : public PeriodicTask {
+  const double _percent;
+  size_t _limit;
+
+  void update_limit() {
+    const size_t limit_100 = os::physical_memory();
```

Comment:
Can this change dynamically?

Reply:
Yes; os::physical_memory() is, when run in a container, fed from the dynamically updated container limit.

There is an existing issue (I think at least it is one): reading limits is expensive since we read the proc file system. But container limits only change very rarely. At the moment, we have a timeout for container data, so we don't read the limits every time we access `os::physical_memory(),` but the timeout is hard coded and very short. Some milliseconds. I think this timeout should be a lot larger and maybe be configurable. Have not opened a JBS issue yet since I wanted feedback first from @jerboaa .

Reply:
> This timeout should be a lot larger and possibly be configurable. Have not opened a JBS issue yet since I wanted feedback first from @jerboaa .

There was some discussion around making the cache timeout configurable at some point but wasn't done, because it was not clear it was needed. See https://bugs.openjdk.org/browse/JDK-8296125. If there are more consumers and periodic polls of `os::physical_memory()` then it becomes more compelling to make this tunable (I haven't really looked at the patch yet). At the time, using `-XX:-UseDynamicNumberOfCompilerThreads` was sufficient.